### PR TITLE
SageMakerLLMEndPoint

### PR DESCRIPTION
**Pull request by iron8kid**
# Description

Adding SageMakerLLMEndPoint to be able to interact with LLM deployed in SageMaker endpoints.

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update (adding it to llms list)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings

### CODE REVIEW

File: llama_index/llms/sagemaker_llm_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                yield message
+
+        # NOTE: convert generator to async generator
+        return gen()
+
+    @llm_completion_callback()
+    async def acomplete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponse:
+        return self.complete(prompt, formatted=formatted, **kwargs)
```
Comment (nerdai) [inline llama_index/llms/sagemaker_llm_endpoint.py:259]: similar to `Bedrock` LLM, can we just `raise NotImplementedError` here?

File: llama_index/llms/sagemaker_llm_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    @llm_completion_callback()
+    async def astream_complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponseAsyncGen:
+        async def gen() -> CompletionResponseAsyncGen:
+            for message in self.stream_complete(prompt, formatted=formatted, **kwargs):
+                yield message
+
+        # NOTE: convert generator to async generator
+        return gen()
```
Comment (nerdai) [inline llama_index/llms/sagemaker_llm_endpoint.py:270]: similar to `Bedrock` LLM, can we just `raise NotImplementedError` here?

File: llama_index/llms/sagemaker_llm_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        self,
+        messages: Sequence[ChatMessage],
+        **kwargs: Any,
+    ) -> ChatResponseAsyncGen:
+        async def gen() -> ChatResponseAsyncGen:
+            for message in self.stream_chat(messages, **kwargs):
+                yield message
+
+        # NOTE: convert generator to async generator
+        return gen()
```
Comment (nerdai) [inline llama_index/llms/sagemaker_llm_endpoint.py:253]: similar to `Bedrock` LLM, can we just `raise NotImplementedError` here?

File: llama_index/llms/sagemaker_llm_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        completion_response_gen = self.stream_complete(prompt, formatted=True, **kwargs)
+        return stream_completion_response_to_chat_response(completion_response_gen)
+
+    @llm_chat_callback()
+    async def achat(
+        self,
+        messages: Sequence[ChatMessage],
+        **kwargs: Any,
+    ) -> ChatResponse:
+        return self.chat(messages, **kwargs)
```
Comment (nerdai) [inline llama_index/llms/sagemaker_llm_endpoint.py:240]: similar to `Bedrock` LLM, can we just `raise NotImplementedError` here?

File: llama_index/llms/sagemaker_llm_endpoint.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        async def gen() -> CompletionResponseAsyncGen:
+            for message in self.stream_complete(prompt, formatted=formatted, **kwargs):
+                yield message
+
+        # NOTE: convert generator to async generator
+        return gen()
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "sagemaker_endpoint"
```
Comment (nerdai) [inline llama_index/llms/sagemaker_llm_endpoint.py:274]: nit: can we do "sagemaker_llm_endpoint"?

**Review by nerdai — APPROVED** [review]
Thanks for this @iron8kid! Left some minor comments.

**Comment by nerdai** [comment]
Looks like we need to get lint/format check to pass still for python 3.9. Can you check the details and resolve the listed issues?

File: llama_index/llms/sagemaker_llm_endpoint.py (hunk 1 above)
Comment (iron8kid) [reply llama_index/llms/sagemaker_llm_endpoint.py]: corrected!

File: llama_index/llms/sagemaker_llm_endpoint.py (hunk 5 above)
Comment (iron8kid) [reply llama_index/llms/sagemaker_llm_endpoint.py]: corrected!

File: llama_index/llms/sagemaker_llm_endpoint.py (hunk 2 above)
Comment (iron8kid) [reply llama_index/llms/sagemaker_llm_endpoint.py]: corrected!

**Review by nerdai — APPROVED** [review]
Thanks for making the changes!

**Comment by iron8kid** [comment]
> Thanks for this @iron8kid! Left some minor comments.

rectification done! 
In the next PR will be adding SageMakerEmbedEndpoint ;)

**Comment by nerdai** [comment]
merging now :)

**Pull request merged**